File: src/com/intellij/idea/plugin/hybris/type/system/structure/view/TSStructureTreeElement.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2016 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.type.system.structure.view;

import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.idea.plugin.hybris.common.utils.HybrisIcons;
import com.intellij.idea.plugin.hybris.type.system.model.Attribute;
import com.intellij.idea.plugin.hybris.type.system.model.CollectionType;
import com.intellij.idea.plugin.hybris.type.system.model.ColumnType;
import com.intellij.idea.plugin.hybris.type.system.model.CustomProperty;
import com.intellij.idea.plugin.hybris.type.system.model.Deployment;
import com.intellij.idea.plugin.hybris.type.system.model.Description;
import com.intellij.idea.plugin.hybris.type.system.model.EnumType;
import com.intellij.idea.plugin.hybris.type.system.model.EnumValue;
import com.intellij.idea.plugin.hybris.type.system.model.Index;
import com.intellij.idea.plugin.hybris.type.system.model.ItemType;
import com.intellij.idea.plugin.hybris.type.system.model.MapType;
import com.intellij.idea.plugin.hybris.type.system.model.Modifiers;
import com.intellij.idea.plugin.hybris.type.system.model.Persistence;
import com.intellij.idea.plugin.hybris.type.system.model.Relation;
import com.intellij.idea.plugin.hybris.type.system.model.RelationElement;
import com.intellij.idea.plugin.hybris.type.system.model.Type;
import com.intellij.idea.plugin.hybris.type.system.model.TypeGroup;
import com.intellij.idea.plugin.hybris.type.system.model.Value;
import com.intellij.psi.xml.XmlElement;
import com.intellij.psi.xml.XmlTag;
import com.intellij.psi.xml.XmlTagValue;
import com.intellij.util.Function;
import com.intellij.util.xml.DomElement;
import com.intellij.util.xml.DomElementNavigationProvider;
import com.intellij.util.xml.DomService;
import com.intellij.util.xml.GenericAttributeValue;
import com.intellij.util.xml.structure.DomStructureTreeElement;
import org.apache.commons.lang3.BooleanUtils;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.List;

import static com.intellij.util.containers.ContainerUtilRt.newArrayList;

/**
 * Created by Martin Zdarsky-Jones (martin.zdarsky@hybris.com) on 20/2/17.
 */
public class TSStructureTreeElement extends DomStructureTreeElement {

    private final Function<DomElement, DomService.StructureViewMode> myDescriptor;
    private final DomElementNavigationProvider myNavigationProvider;

    public <T extends DomElement> TSStructureTreeElement(
        final T stableCopy,
        final Function<DomElement, DomService.StructureViewMode> myDescriptor,
        final DomElementNavigationProvider myNavigationProvider
    ) {
        super(stableCopy, myDescriptor, myNavigationProvider);
        this.myDescriptor = myDescriptor;
        this.myNavigationProvider = myNavigationProvider;
    }

    protected StructureViewTreeElement createChildElement(final DomElement element) {
        return new TSStructureTreeElement(element, myDescriptor, myNavigationProvider);
    }

    /**
     * Returns the name of the object to be presented in most renderers across the program.
     *
     * @return the object name.
     */
    @Nullable
    public String getPresentableText() {
        final DomElement dom = getElement();
        if (dom instanceof Attribute) {
            final GenericAttributeValue<String> qualifier = ((Attribute) dom).getQualifier();
            String value = resolveValue(qualifier);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof CollectionType) {
            final GenericAttributeValue<String> code = ((CollectionType) dom).getCode();
            String value = resolveValue(code);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof ColumnType) {
            final GenericAttributeValue<String> database = ((ColumnType) dom).getDatabase();
            String value = resolveValue(database);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof CustomProperty) {
            final GenericAttributeValue<String> name = ((CustomProperty) dom).getName();
            String value = resolveValue(name);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof Deployment) {
            final GenericAttributeValue<String> table = ((Deployment) dom).getTable();
            String value = resolveValue(table);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof EnumType) {
            final GenericAttributeValue<String> code = ((EnumType) dom).getCode();
            String value = resolveValue(code);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof EnumValue) {
            final GenericAttributeValue<String> code = ((EnumValue) dom).getCode();
            String value = resolveValue(code);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof Index) {
            final GenericAttributeValue<String> name = ((Index) dom).getName();
            String value = resolveValue(name);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof ItemType) {
            final GenericAttributeValue<String> code = ((ItemType) dom).getCode();
            String value = resolveValue(code);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof MapType) {
            final GenericAttributeValue<String> code = ((MapType) dom).getCode();
            String value = resolveValue(code);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof Persistence) {
            final GenericAttributeValue<Type> type = ((Persistence) dom).getType();
            if (type != null) {
                final String value = type.getStringValue();
                if (value != null) {
                    return value;
                }
            }
        }
        if (dom instanceof Relation) {
            final GenericAttributeValue<String> code = ((Relation) dom).getCode();
            String value = resolveValue(code);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof RelationElement) {
            final GenericAttributeValue<String> qualifier = ((RelationElement) dom).getQualifier();
            String value = resolveValue(qualifier);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof TypeGroup) {
            final GenericAttributeValue<String> name = ((TypeGroup) dom).getName();
            String value = resolveValue(name);
            if (value != null) {
                return value;
            }
        }
        if (dom instanceof Value) {
            final GenericAttributeValue<String> code = ((Value) dom).getCode();
            String value = resolveValue(code);
            if (value != null) {
                return value;
            }
        }
        return super.getPresentableText();
    }

    private String resolveValue(final GenericAttributeValue<String> attributeValue) {
        if (attributeValue != null) {
            return attributeValue.getStringValue();
        }
        return null;
    }

    /**
     * Returns the location of the object (for example, the package of a class). The location
     * string is used by some renderers and usually displayed as grayed text next to the item name.
     *
     * @return the location description, or null if none is applicable.
     */
    @Nullable
    public String getLocationString() {
        final DomElement dom = getElement();
        final String attrValue = localizeAttribute(dom);
        if (attrValue != null) {
            return attrValue;
        }
        final String itemTypeValue = localizeItemType(dom);
        if (itemTypeValue != null) {
            return itemTypeValue;
        }
        final String modifiersValue = localizeModifiers(dom);
        if (modifiersValue != null) {
            return modifiersValue;
        }
        final String descriptionValue = localizeDescription(dom);
        if (descriptionValue != null) {
            return descriptionValue;
        }
        final String xmlElementName = dom.getXmlElementName();
        if (xmlElementName == null) {
            return null;
        }

        if (xmlElementName.equalsIgnoreCase(getPresentableText())) {
            return null;
        }
        return xmlElementName;
    }

    private String localizeDescription(final DomElement dom) {
        if (dom instanceof Description) {
            final XmlElement xmlElement = dom.getXmlElement();
            if (xmlElement instanceof XmlTag) {
                final XmlTagValue value = ((XmlTag) xmlElement).getValue();
                if (value != null) {
                    return value.getTrimmedText();
                }
            }
        }
        return null;
    }

    private String localizeModifiers(final DomElement dom) {
        if (dom instanceof Modifiers) {
            final Modifiers modifiers = (Modifiers) dom;
            final List<String> modList = newArrayList();
            resolveModifier(modifiers.getDoNotOptimize(), "doNotOptimize", modList);
            resolveModifier(modifiers.getEncrypted(), "encrypted", modList);
            resolveModifier(modifiers.getInitial(), "initial", modList);
            resolveModifier(modifiers.getOptional(), "optional", modList);
            resolveModifier(modifiers.getPartOf(), "partOf", modList);
            resolveModifier(modifiers.getPrivate(), "private", modList);
            resolveModifier(modifiers.getRead(), "read", modList);
            resolveModifier(modifiers.getRemovable(), "removable", modList);
            resolveModifier(modifiers.getSearch(), "search", modList);
            resolveModifier(modifiers.getUnique(), "unique", modList);
            resolveModifier(modifiers.getWrite(), "write", modList);
            if (!modList.isEmpty()) {
                return StringUtils.join(modList, ", ");
            }
        }
        return null;
    }

    private void resolveModifier(
        final GenericAttributeValue<Boolean> attribute,
        final String methodName,
        final List<String> resultList
    ) {
        if (attribute == null) {
            return;
        }
        final Boolean value = BooleanUtils.toBooleanObject(attribute.getStringValue());
        if (value == null) {
            return;
        }
        if (value) {
            resultList.add(methodName);
        } else {
            resultList.add(methodName + "(false)");
        }

    }

    private String localizeItemType(final DomElement dom) {
        if (dom instanceof ItemType) {
            ItemType itemType = (ItemType) dom;
            String value = resolveValue(itemType.getExtends());
            if (value == null) {
                value = "GenericItem";
            }
            return "(" + value + ")";
        }
        return null;
    }

    private String localizeAttribute(final DomElement dom) {
        if (dom instanceof Attribute) {
            final Attribute attribute = (Attribute) dom;
            String value = resolveValue(attribute.getType());
            if (value == null) {
                return null;
            }
            if (value.startsWith("localized:")) {
                value = value.substring("localized:".length());
            }
            if (value.startsWith("java.lang.")) {
                value = value.substring("java.lang.".length());
            }
            return value;
        }
        return null;
    }

    @Override
    @Nullable
    public Icon getIcon(boolean open) {
        final DomElement dom = getElement();
        if (dom instanceof Attribute) {
            final String value = resolveValue(((Attribute) dom).getType());
            if (StringUtils.startsWith(value, "localized:")) {
                return HybrisIcons.LOCALIZED;
            }
        }
        return null;
    }

}

File: src/com/intellij/idea/plugin/hybris/structureView/TSStructureTreeElement.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019 EPAM Systems <hybrisideaplugin@epam.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.intellij.idea.plugin.hybris.structureView

import com.intellij.icons.AllIcons
import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.idea.plugin.hybris.common.HybrisConstants
import com.intellij.idea.plugin.hybris.common.HybrisConstants.TS_ATTRIBUTE_LOCALIZED_PREFIX
import com.intellij.idea.plugin.hybris.common.utils.HybrisIcons
import com.intellij.idea.plugin.hybris.system.type.model.*
import com.intellij.psi.xml.XmlTag
import com.intellij.util.Function
import com.intellij.util.xml.DomElement
import com.intellij.util.xml.DomElementNavigationProvider
import com.intellij.util.xml.DomService
import com.intellij.util.xml.GenericAttributeValue
import com.intellij.util.xml.structure.DomStructureTreeElement

class TSStructureTreeElement(
    stableCopy: DomElement,
    private val myDescriptor: Function<DomElement, DomService.StructureViewMode>,
    private val myNavigationProvider: DomElementNavigationProvider
) : DomStructureTreeElement(stableCopy, myDescriptor, myNavigationProvider) {
    override fun createChildElement(element: DomElement): StructureViewTreeElement {
        return TSStructureTreeElement(element, myDescriptor, myNavigationProvider)
    }

    override fun getPresentableText() = when (val dom = element) {
        is AtomicTypes -> "Atomic types"
        is CollectionTypes -> "Collections types"
        is MapTypes -> "Map types"
        is EnumTypes -> "Enum types"
        is Relations -> "Relations"
        is ItemTypes -> "Item types"
        is TypeGroup -> resolveValue(dom.name)
        is AtomicType -> resolveValue(dom.clazz)
        is Attribute -> resolveValue(dom.qualifier)
        is CollectionType -> resolveValue(dom.code)
        is ColumnType -> resolveValue(dom.database)
        is CustomProperty -> resolveValue(dom.name)
        is EnumType -> resolveValue(dom.code)
        is EnumValue -> resolveValue(dom.code)
        is Index -> resolveValue(dom.name)
        is IndexKey -> resolveValue(dom.attribute)
        is IndexInclude -> resolveValue(dom.attribute)
        is ItemType -> resolveValue(dom.code)
        is MapType -> resolveValue(dom.code)
        is Relation -> resolveValue(dom.code)
        is RelationElement -> resolveValue(dom.qualifier)
        is Value -> resolveValue(dom.code)
        is Modifiers -> "modifiers: "
        is Description -> ""
        else -> null
    }
        ?: super.getPresentableText()

    override fun getLocationString() = when (val dom = element) {
        is Attribute -> resolveLocationString(dom)
        is ItemType -> resolveLocationString(dom)
        is Modifiers -> resolveLocationString(dom)
        is Description -> resolveLocationString(dom)
        is Index -> resolveLocationString(dom)
        is CollectionType -> resolveLocationString(dom)
        is MapType -> resolveLocationString(dom)
        is RelationElement -> resolveLocationString(dom)
        is ModelMethod -> resolveLocationString(dom)
        is Deployment -> resolveLocationString(dom)
        is Persistence -> dom.type.stringValue + (dom.attributeHandler.stringValue?.let { " ($it)" } ?: "")
        is EnumType -> dom.dynamic.value?.let { "dynamic" }
        is Relation -> dom.localized.value?.let { "localized" }
        is AtomicType -> dom.extends.stringValue
        is CustomProperty -> dom.value.stringValue
        is IndexKey -> "key"
        is IndexInclude -> "include"
        else -> null
    }

    private fun resolveValue(attributeValue: GenericAttributeValue<String>?) = attributeValue?.stringValue

    private fun resolveLocationString(dom: MapType) = (dom.argumentType.stringValue ?: "?") + " : " + (dom.returnType.stringValue ?: "?")

    private fun resolveLocationString(dom: CollectionType) =
        (dom.type.stringValue ?: "collection") + (dom.elementType.stringValue?.let { " of $it" } ?: "")

    private fun resolveLocationString(dom: Description): String? {
        val xmlElement = dom.xmlElement
        if (xmlElement is XmlTag) {
            return xmlElement.value.trimmedText
        }
        return null
    }

    private fun resolveLocationString(dom: Deployment) = listOfNotNull(
        dom.table.stringValue,
        dom.typeCode.stringValue?.let { "($it)" },
    ).joinToString("")

    private fun resolveLocationString(dom: ModelMethod) = listOfNotNull(
        dom.name.stringValue,
        dom.default.value?.let { "default($it)" },
        dom.deprecated.value?.let { "deprecated($it)" },
        dom.deprecatedSince.stringValue?.let { "since $it" }
    ).joinToString()

    private fun resolveLocationString(dom: RelationElement) = listOfNotNull(
        dom.type.stringValue,
        dom.collectionType.stringValue ?: "collection",
        dom.cardinality.stringValue,
        dom.ordered.stringValue?.let { "ordered($it)" }
    ).joinToString()

    private fun resolveLocationString(dom: Index) = listOfNotNull(
        dom.creationMode.value?.let { "mode(${it.name})" },
        dom.remove.value?.let { "${Index.REMOVE}($it)" },
        dom.replace.value?.let { "${Index.REPLACE}($it)" },
        dom.unique.value?.let { "${Index.UNIQUE}($it)" }
    ).joinToString()

    private fun resolveLocationString(dom: Modifiers) = listOfNotNull(
        dom.doNotOptimize.value?.let { "${Modifiers.DONT_OPTIMIZE}($it)" },
        dom.encrypted.value?.let { "${Modifiers.ENCRYPTED}($it)" },
        dom.initial.value?.let { "${Modifiers.INITIAL}($it)" },
        dom.optional.value?.let { "${Modifiers.OPTIONAL}($it)" },
        dom.partOf.value?.let { "${Modifiers.PART_OF}($it)" },
        dom.private.value?.let { "${Modifiers.PRIVATE}($it)" },
        dom.read.value?.let { "${Modifiers.READ}($it)" },
        dom.removable.value?.let { "${Modifiers.REMOVABLE}($it)" },
        dom.search.value?.let { "${Modifiers.SEARCH}($it)" },
        dom.unique.value?.let { "${Modifiers.UNIQUE}($it)" },
        dom.write.value?.let { "${Modifiers.WRITE}($it)" }
    ).joinToString()

    private fun resolveLocationString(dom: ItemType) = resolveValue(dom.extends)?.let { ": $it" } ?: ": ${HybrisConstants.TS_TYPE_GENERIC_ITEM}"

    private fun resolveLocationString(dom: Attribute): String? {
        var value = resolveValue(dom.type) ?: return null
        if (value.startsWith(TS_ATTRIBUTE_LOCALIZED_PREFIX)) {
            value = value.substring(TS_ATTRIBUTE_LOCALIZED_PREFIX.length)
            value += " (localized)"
        }
        if (value.startsWith("java.lang.")) {
            value = value.substring("java.lang.".length)
        }
        return value
    }

    override fun getIcon(open: Boolean) = when (element) {
        is Description -> AllIcons.Windows.Help
        is AtomicType -> HybrisIcons.ATOMIC
        is EnumType -> HybrisIcons.ENUM
        is EnumValue -> HybrisIcons.ENUM_VALUE
        is Attribute -> HybrisIcons.ATTRIBUTE
        is Index -> HybrisIcons.INDEX
        is ItemType -> HybrisIcons.ITEM
        is CollectionType -> HybrisIcons.COLLECTION
        is CustomProperty -> HybrisIcons.PROPERTY
        is MapType -> HybrisIcons.MAP
        is Relation -> HybrisIcons.RELATION
        is RelationSourceElement -> HybrisIcons.RELATION_SOURCE
        is RelationTargetElement -> HybrisIcons.RELATION_TARGET
        else -> null
    }

//    : Icon? {
//        val dom = element
//        // TODO: add icons
//        if (dom is Attribute) {
//            val value = resolveValue(dom.type)
//            if (StringUtils.startsWith(value, TS_ATTRIBUTE_LOCALIZED_PREFIX)) {
//                return HybrisIcons.LOCALIZED
//            }
//        }
//        return null
//    }
}